Assembler.UnitTests/InitialMessageHandlerTests.cs

```diff
@@ -0,0 +1,182 @@
+
+            // Assert
+            _identifierFactoryMock.Verify(identifier => identifier.Create(It.IsAny<BaseFrame>()), Times.Once);
+            _identifierFactoryMock.Verify(identifier => identifier.Create(frame.Object), Times.Once);
+        }
+
+        [Test]
+        public void Handle_MessageInCacheAndMiddleReceived_ReleasesTheOldOneAndCreatesANewOne()
+        {
+            // Arrange
+            var frame = new Mock<BaseFrame>(FrameType.Initial);
+            var message = new Mock<BaseMessageInAssembly>();
+            var newMessage = new Mock<BaseMessageInAssembly>();
+            message.Object.MiddleReceived = true;
+
+            _cacheMock.Setup(cache => cache.Exists(It.IsAny<string>())).Returns(true);
+            _cacheMock.Setup(cache => cache.Get<BaseMessageInAssembly>(It.IsAny<string>())).Returns(message.Object);
+            _messageInAssemblyCreatorMock.Setup(creator => creator.Create()).Returns(newMessage.Object);
+
+            // Act
+            _handler.Handle(frame.Object);
+
+            // Assert
+            _identifierFactoryMock.Verify(identifier => identifier.Create(It.IsAny<BaseFrame>()), Times.Once);
+            _identifierFactoryMock.Verify(identifier => identifier.Create(frame.Object), Times.Once);
+
+            _cacheMock.Verify(cache => cache.Exists(It.IsAny<string>()), Times.Once);
+            _cacheMock.Verify(cache => cache.Exists(_identifierString), Times.Once);
+
+            _cacheMock.Verify(cache => cache.Get<It.IsAnyType>(It.IsAny<string>()), Times.Once);
+            _cacheMock.Verify(cache => cache.Get<It.IsAnyType>(_identifierString), Times.Once);
+
+            _cacheMock.Verify(cache => cache.Remove(It.IsAny<string>()), Times.Once);
+            _cacheMock.Verify(cache => cache.Remove(_identifierString), Times.Once);
+
+            Assert.AreEqual(ReleaseReason.AnotherMessageStarted, message.Object.ReleaseReason);
+
+            Assert.AreEqual(1, _assembledMessages.Count);
+            Assert.AreEqual(message.Object, _assembledMessages.First());
+
+            _messageInAssemblyCreatorMock.Verify(creator => creator.Create(), Times.Once);
+
+            _enricherMock.Verify(enricher => enricher.Enrich(It.IsAny<BaseFrame>(), It.IsAny<BaseMessageInAssembly>()),
+                Times.Once);
+            _enricherMock.Verify(enricher => enricher.Enrich(frame.Object, newMessage.Object), Times.Once);
+
+            _cacheMock.Verify(cache => cache.Put(It.IsAny<string>(), It.IsAny<BaseMessageInAssembly>()), Times.Once);
+            _cacheMock.Verify(cache => cache.Put(_identifierString, newMessage.Object), Times.Once);
+        }
+
+        [Test]
+        public void Handle_MessageInCacheAndMiddleNotReceived_EnrichesTheExistingMessage()
+        {
+            // Arrange
+            var frame = new Mock<BaseFrame>(FrameType.Initial);
+            var message = new Mock<BaseMessageInAssembly>();
+            message.Object.MiddleReceived = false;
```

Comment:
`message.SetupGet(msg => msg.MiddleReceived).Returns(false);`

Reply:
Throws exception, I used it before but it wouldn't let me.

Reply:
See https://github.com/reverser-gever/Assembler/pull/2#discussion_r397139144